src/components/views/rooms/RoomTile.tsx

```diff
@@ -81,182 +60,104 @@ const contextMenuBelow = (elementRect: PartialDOMRect) => {
+            this.model.on("change", this.rerender);
+            this.registerActions(this.model);
+        }
+    }
 
-    private onLocalEchoUpdated = (ev: MatrixEvent, room: Room) => {
-        if (!room?.roomId === this.props.room.roomId) return;
-        this.setState({hasUnsentEvents: this.countUnsentEvents() > 0});
-    };
+    public componentWillUnmount() {
+        this.model.off("change", this.rerender);
+        this.unregisterActions(this.model);
+        this.model.destroy();
+    }
 
-    private onRoomPropertyUpdate = (property: CachedRoomKey) => {
-        if (property === CachedRoomKey.NotificationVolume) this.onNotificationUpdate();
-        // else ignore - not important for this tile
+    private rerender = () => {
+        this.forceUpdate();
     };
 
-    private get showContextMenu(): boolean {
-        return this.props.tag !== DefaultTagID.Invite;
+    private registerActions(model: V) {
+        if (this.actions) {
+            for (const [name, handler] of Object.entries(this.actions)) {
+                this.model.on(name, handler);
+            }
+        }
     }
 
-    private get showMessagePreview(): boolean {
-        return !this.props.isMinimized && this.props.showMessagePreview;
+    private unregisterActions(model: V) {
+        if (this.actions) {
+            for (const [name, handler] of Object.entries(this.actions)) {
+                this.model.off(name, handler);
+            }
+        }        
     }
+}
 
-    public componentDidUpdate(prevProps: Readonly<IProps>, prevState: Readonly<IState>) {
-        if (prevProps.showMessagePreview !== this.props.showMessagePreview && this.showMessagePreview) {
-            this.setState({messagePreview: this.generatePreview()});
-        }
-        if (prevProps.room?.roomId !== this.props.room?.roomId) {
-            MessagePreviewStore.instance.off(
-                MessagePreviewStore.getPreviewChangedEventName(prevProps.room),
-                this.onRoomPreviewChanged,
-            );
-            MessagePreviewStore.instance.on(
-                MessagePreviewStore.getPreviewChangedEventName(this.props.room),
-                this.onRoomPreviewChanged,
-            );
-            CommunityPrototypeStore.instance.off(
-                CommunityPrototypeStore.getUpdateEventName(prevProps.room?.roomId),
-                this.onCommunityUpdate,
-            );
-            CommunityPrototypeStore.instance.on(
-                CommunityPrototypeStore.getUpdateEventName(this.props.room?.roomId),
-                this.onCommunityUpdate,
-            );
-            prevProps.room?.off("Room.name", this.onRoomNameUpdate);
-            this.props.room?.on("Room.name", this.onRoomNameUpdate);
-        }
+@replaceableComponent("views.rooms.RoomTile")
+export default class RoomTile extends ModelView<RoomTileViewModel, IState> {
+    private roomTileRef = createRef<HTMLDivElement>();
+
+    constructor(props: IProps<RoomTileViewModel>) {
```

Comment:
is there a way to declare IProps on the ModelView class already taking the passed in view model type into account?
E.g. something like:
```ts
type Props = IProps<V>
```
in ModelView and then here just do `RoomTile.Props` instead of `IProps<RoomTileViewModel>` or something?
This probably doesn't make any sense, haven't done typescript in a while...

Just so I would have to export IProps if I were to extract ModelView to its own file.

Reply:
I'm not sure I'm grokking exactly what you're asking but I think you're asking about `ComponentProps<typeof XYZ>` (React.ComponentProps)

Reply:
Interesting, could you give an example of how this could be wrapped?

Reply:
actually I think I messed that up or have since lost it, /me looks at @turt2live for inspiration

Reply:
most IDEs will infer the types for people if you just leave it off the constructor, so:

```
constructor(props) {
  super(props); // it now knows that props is IProps<T>
}

Reply:
also, imo, given we don't use `props` but instead `this.props` in constructors, I'm not too worried about types - we're just fulfilling the contract, which doesn't need explicit type declarations.

The other option would be to make the props the view model, so people can `<Component {...model} />`